### PR TITLE
Fix `bundle install` on truffleruby selecting incorrect variant for `sorbet-static` gem

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
## What was the end-user or developer problem that led to this PR?

Bundler selecting incorrect variants on truffleruby.

## What is your fix for the problem, implemented in this PR?

Make sure we go through custom truffleruby logic: https://github.com/oracle/truffleruby/blob/72ce69ec079618a3c0513da968262a4484633834/lib/mri/rubygems/platform.rb#L44.

Fixes https://github.com/rubygems/rubygems/issues/4588.

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

**Comment by eregon** [comment]
@deivid-rodriguez Do you need any help with this PR, anything I can do?

From a quick look at the CI there is some lint errors, and some other failures which I'm not sure are related to changes here.

File: bundler/lib/bundler/rubygems_ext.rb [hunk 1]
```diff
@@ -174,20 +174,36 @@ def hash
...
     MINGW = Gem::Platform.new("x86-mingw32") unless defined?(MINGW)
     X64_MINGW = Gem::Platform.new("x64-mingw32") unless defined?(X64_MINGW)
+  end
 
-    undef_method :hash if method_defined? :hash
-    def hash
-      @cpu.hash ^ @os.hash ^ @version.hash
-    end
+  Platform.singleton_class.module_eval do
+    unless Platform.singleton_methods.include?(:match_spec?)
```
Comment (eregon) [inline bundler/lib/bundler/rubygems_ext.rb:188]: Indeed, they were all introduced at the same time in https://github.com/rubygems/rubygems/pull/3817/files#diff-b2eac061756123f15f8d8e41075e4b8466a60945cad4e594b4eadf241d1a4089
So to backfill them it sounds like indeed all 3 need to be backfilled.

**Comment by deivid-rodriguez** [comment]
Last time I looked they seemed related, I'll try to find some time to look at it but you can have a look yourself if you want too.

**Comment by deivid-rodriguez** [comment]
@eregon I had another look and I think I was wrong and the problem was indeed unrelated to this PR. And in any case, I think it should've been workarounded by #4736. So I rebased this and let's see how it goes.

**Comment by eregon** [comment]
Thanks :)

**Pull request merged**